### DISCUSSION TITLE
Chemaxon Extended SMILES and SMARTS for chemical classes definitions

### DISCUSSION THREAD

**Discussion by Adafede** [rdkit/rdkit#6766] [Q&A]
Dear RDKit community (and even beyond regarding the `CX` discussion),

For a bit of context, I am trying to define _chemical classes_ in the lesss ambiguous possible way.
I already went through the two blog posts [searching-with-generic-groups](https://greglandrum.github.io/rdkit-blog/posts/2022-04-05-searching-with-generic-groups.html) and [generalized-substructure-search](https://greglandrum.github.io/rdkit-blog/posts/2021-08-03-generalized-substructure-search.html).

I would like to share my thoughts about some of the examples and extensions in the following thread.
It encompasses different questions, and I am very happy to open distinct issues for each of the points relevant to the whole community and/or summarize it as a blog post later on.

## About `AdjustQueryParameters`

Following [searching-with-generic-groups](https://greglandrum.github.io/rdkit-blog/posts/2022-04-05-searching-with-generic-groups.html), I took the following structure (`structure 1`).

### structure 1

![image](https://github.com/rdkit/rdkit/assets/44283913/5e018f17-1b37-4e8a-bb2f-47011abf92b5)
```plaintext
*-[c]1[n][c]2[c][c][c][n][c]2[n]1 |$CAL_p;;;;;;;;;$|
```

I would consider it as _ambiguous_ as already highlighted in the blog post as it requires to fine tune the `AdjustQueryParameters`.

A way to be clearer would be to define the number of explicit bonds, as in `structure 2` and `structure 3`:

### structure 2

![image](https://github.com/rdkit/rdkit/assets/44283913/f03a89f5-2753-451a-b1f8-a35ee5ab6b3c)
```plaintext
*-[c;D3]1[n;D2,D3][c;D3]2[c;D2,D3][c;D2,D3][c;D2,D3][n;D2,D3][c;D3]2[n;D2,D3]1 |$CAL_p;;;;;;;;;$|
```

### structure 3

![image](https://github.com/rdkit/rdkit/assets/44283913/a5e68dce-024a-4dbf-a195-4a20b9c00524)
```plaintext
*-[c;D3]1[n;D2][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n;D2]1 |$CAL_p;;;;;;;;;$|
```

However, when doing so, if using `AdjustQueryParameters` afterwards, an additional layer of `D...` is added, even if the information is already present. Shouldn't the default behaviour check first, and only add it if needed?

In detail:

```python
from rdkit import Chem


# Examples from the discussion
queries = [
    ("Structure 1",
     "*-[c]1[n][c]2[c][c][c][n][c]2[n]1 |$CAL_p;;;;;;;;;$|"),
    ("Structure 2",
     "*-[c;D3]1[n;D2,D3][c;D3]2[c;D2,D3][c;D2,D3][c;D2,D3][n;D2,D3][c;D3]2[n;D2,D3]1 |$CAL_p;;;;;;;;;$|"),
    ("Structure 3",
     "*-[c;D3]1[n;D2][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n;D2]1 |$CAL_p;;;;;;;;;$|"),
]

for query_id, query_structure in queries:
    q = Chem.MolFromSmarts(query_structure)
    print(Chem.MolToCXSmarts(q))
    q2 = Chem.AdjustQueryProperties(q)
    print(Chem.MolToCXSmarts(q2))
```

## About variable attachment points

My second thoughts are for the variable attachment points described in [generalized-substructure-search](https://greglandrum.github.io/rdkit-blog/posts/2021-08-03-generalized-substructure-search.html).

Starting from `structure 4`:

### structure 4

![image](https://github.com/rdkit/rdkit/assets/44283913/5d961e66-a026-422f-bfc2-23b1893e5204)
```plaintext
[#6]-[*].[c;D2]1[n][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n]1 |m:1:10.3|
```

How could one write down the following instructions in CXSMARTS:

- Only one `[#6]` can be attached, and nothing else.
- One, or two `[#6]` can be attached, but nothing else.
- Only one `[#6]` can be attached, and if something else is attached on the other attachment point, it **cannot** be `[#6]`.

## About [R-logic](https://docs.chemaxon.com/display/docs/chemaxon-extended-smiles-and-smarts-cxsmiles-and-cxsmarts.md#src-1806633-chemaxonextendedsmilesandsmarts-cxsmilesandcxsmarts-r-logic)

Trying to answer the above question, I went back through the [CXSMILES/SMARTS documentation](https://docs.chemaxon.com/display/docs/chemaxon-extended-smiles-and-smarts-cxsmiles-and-cxsmarts.md) and realized the variable attachment points eventually were not the best approach, and one should instead use `R-groups`, complemented with some [R-logic](https://docs.chemaxon.com/display/docs/chemaxon-extended-smiles-and-smarts-cxsmiles-and-cxsmarts.md#src-1806633-chemaxonextendedsmilesandsmarts-cxsmilesandcxsmarts-r-logic).

I then went to <https://www.rdkit.org/docs/RDKit_Book.html#cxsmiles-cxsmarts-extensions> and could not see the logic being supported.
So my first question is: Are they supported now but not documented or still not?

In any case, could they be theoretically used to express some more complicated things as in the following:

Taking this time a flavonoid looking like `structure 5`:

### structure 5

![image](https://github.com/rdkit/rdkit/assets/44283913/cf3993b4-40cb-4ef4-9844-cc1dbe7f66e3)
```plaintext
[#8;D1]-[*].[#8;D1]-[*].O=[#6]-1-[#6]-[#6](-[#8]-c2ccccc-12)-c1ccccc1 |m:1:10.11.12.13.16.17.18.19.20,3:10.11.12.13.16.17.18.19.20|
```

How could one write down the following instructions in CXSMARTS:

- Only one `[#8;D1]` can be attached on both rings, and the rest **cannot** be `[#8;D1]`.

I started thinking about it and I thought about something like `structure 6` (on one ring only for simplicity):

### structure 6

![image](https://github.com/rdkit/rdkit/assets/44283913/6968ba03-6874-49b0-891a-5195291c41db)
```plaintext
[*]-[c]1[c](-[*])[c](-[*])[c;D3]2-[#6](=[O])-[#6;D2]-[#6;D3](-[#8;D2]-[c;D3]2[c]1-[*])-[c;D3]1[c][c][c][c][c]1 |$_R1;;;_R2;;_R3;;;;;;;;;_R4;;;;;;$,RG:_R1={[#8;D1]},_R2={[#8;D1]},_R3={[#8;D1]},_R4={[#8;D1]},LOG={_R1:;;0-1._R2:;;0-1._R3:;;0-1._R4:;;0-1}| 
# the logic part is random, I really don't know
```

No idea how one would write down:

- I want R1 + R2 + R3 + R4 = 1. So only one `[#8;D1]` group, rest is free (but no `[#8;D1]`)
- I want R1 + R2 + R3 + R4 = 1. Rest should be empty.
- I want R1 + R2 + R3 + R4 = >1. It can be either 2, 3 or 4.

I hope these questions will open some good discussions, happy to develop any needed point!

## About enumeration with [link nodes](https://docs.chemaxon.com/display/docs/chemaxon-extended-smiles-and-smarts-cxsmiles-and-cxsmarts.md#link-nodes) not starting with `1`

Another personal example I tried to address was the case of `structure 7` that I first wrote as:

### structure 7
![image](https://github.com/rdkit/rdkit/assets/44283913/b8755e04-8b4b-4f4b-ae6b-3f700127a869)
```plaintext
[CH3][CH2][N+0]([CH3])[CH3] |LN:1:11.13|
```

When I tried enumerating it, I saw it was not enumerating what I expected, and so I probably did not write it the way it was expected.
So I tried with `structure 8`, and it worked as expected.

### structure 8
![image](https://github.com/rdkit/rdkit/assets/44283913/e4de0e04-148f-4e1b-bd9a-595a61229f10)
```plaintext
[CH3][CH2][CH2][CH2][CH2][CH2][CH2][CH2][CH2][CH2][CH2][CH2][N+0]([CH3])[CH3] |LN:11:1.3|
```

Should cases with `LN:X:>1` be supported?

## Small bonus

For the ones that would like some bites of code (everything made a lot easier also thanks to @cthoyt and @bjonnh-work):

```python
import chembl_downloader
import os
import pickle
import time

# from chembl_downloader import latest
from pystow import join
from rdkit import Chem

from rdkit.Chem import rdMolEnumerator, rdSubstructLibrary, rdTautomerQuery


# Examples from the discussion
queries = [
    ("Structure 1",
     "*-[c]1[n][c]2[c][c][c][n][c]2[n]1 |$CAL_p;;;;;;;;;$|"),
    ("Structure 2",
     "*-[c;D3]1[n;D2,D3][c;D3]2[c;D2,D3][c;D2,D3][c;D2,D3][n;D2,D3][c;D3]2[n;D2,D3]1 |$CAL_p;;;;;;;;;$|"),
    ("Structure 3",
     "*-[c;D3]1[n;D2][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n;D2]1 |$CAL_p;;;;;;;;;$|"),
    ("Structure 4",
     "[#6]-[*].[c;D2]1[n][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n]1 |m:1:10.3|"),
    ("Structure 5",
     "[#8;D1]-[*].[#8;D1]-[*].O=[#6]-1-[#6]-[#6](-[#8]-c2ccccc-12)-c1ccccc1 |m:1:10.11.12.13.16.17.18.19.20,3:10.11.12.13.16.17.18.19.20|"),
    ("Structure 6",
     "[*]-[c]1[c](-[*])[c](-[*])[c;D3]2-[#6](=[O])-[#6;D2]-[#6;D3](-[#8;D2]-[c;D3]2[c]1-[*])-[c;D3]1[c][c][c][c][c]1 |$_R1;;;_R2;;_R3;;;;;;;;;_R4;;;;;;$,RG:_R1={[#8;D1]},_R2={[#8;D1]},_R3={[#8;D1]},_R4={[#8;D1]},LOG={_R1:;;0-1._R2:;;0-1._R3:;;0-1._R4:;;0-1}|"),
    ("Structure 7",
     "[CH3][CH2][N+0]([CH3])[CH3] |LN:1:11.13|"),
    ("Structure 8",
     "[CH3][CH2][CH2][CH2][CH2][CH2][CH2][CH2][CH2][CH2][CH2][CH2][N+0]([CH3])[CH3] |LN:11:1.3|"),
    # Add more examples if needed
]


def get_latest_chembl():
    try:
        # Get paths and constants
        version, data_path, lib_path = latest_chembl_paths()
        fp_len = 2048
        max_atoms = 50
        report_interval = 50000

        # Create data list
        data = []

        # Download ChEMBL data and process molecules
        with chembl_downloader.supplier(version=version) as suppl:
            t1 = time.time()
            for i, mol in enumerate(suppl):
                if not ((i + 1) % report_interval):
                    print(f"Processed {i + 1} molecules in {(time.time() - t1):.1f} seconds")

                if mol is None or mol.GetNumAtoms() > max_atoms:
                    continue

                fp = Chem.PatternFingerprint(mol, fpSize=fp_len, tautomerFingerprints=True)
                smi = Chem.MolToSmiles(mol)
                data.append((smi, fp))

        # Save data to file
        with open(data_path, 'wb') as file:
            pickle.dump(data, file)

        # Create and save the substructure library
        mols = rdSubstructLibrary.CachedTrustedSmilesMolHolder()
        fps = rdSubstructLibrary.TautomerPatternHolder(fp_len)

        for smi, fp in data:
            mols.AddSmiles(smi)
            fps.AddFingerprint(fp)

        library = rdSubstructLibrary.SubstructLibrary(mols, fps)

        print(f"Processed ChEMBL data in {(time.time() - t1):.2f} seconds. The library has {len(library)} molecules.")

        with open(lib_path, 'wb') as file:
            pickle.dump(library, file)
    except Exception as e:
        print(f"Failed to process ChEMBL data: {str(e)}")


def latest_chembl_paths():
    try:
        # Get the latest ChEMBL version
        # COMMENT AR: For everyone to have the same to compare, fix it to "33".
        # version = latest()
        version = "33"

        # Generate file paths for data and library
        data_path = join("chembl", version, name="sssdata.pkl")
        lib_path = join("chembl", version, name="ssslib.pkl")

        return version, data_path, lib_path
    except Exception as e:
        raise RuntimeError(f"Failed to retrieve ChEMBL paths: {str(e)}")


def load_latest_chembl():
    # Get the latest ChEMBL version and paths
    version, data_path, lib_path = latest_chembl_paths()

    if os.path.exists(lib_path):
        try:
            # Attempt to open and load the SubstructLibrary from the library path
            with open(lib_path, 'rb') as inf:
                sslib = pickle.load(inf)
            print(f'SubstructLibrary loaded with {len(sslib)} molecules')
        except (FileNotFoundError, pickle.PickleError) as e:
            # If there is an issue with loading the library, delete the file and retry
            print(f"Error loading library: {str(e)}")
            os.remove(lib_path)
            print(f"Deleted the corrupted library file. Downloading the latest ChEMBL data...")
            get_latest_chembl()

            # Try to load the library again after downloading
            with open(lib_path, 'rb') as inf:
                sslib = pickle.load(inf)
            print(f'SubstructLibrary loaded with {len(sslib)} molecules')
    else:
        # If the library file is not found, download the latest ChEMBL data
        print("Library file not found. Downloading the latest ChEMBL data...")
        get_latest_chembl()

        # Try to load the library after downloading
        with open(lib_path, 'rb') as inf:
            sslib = pickle.load(inf)
        print(f'SubstructLibrary loaded with {len(sslib)} molecules')

    return sslib

from rdkit import Chem


def search_classes_library(query, sslib, params, tautomer_insensitive=True, max_results=1000):
    """
    Search chemical classes in a library.

    Args:
        query: The defined chemical classes.
        sslib: The chemical library to search.
        tautomer_insensitive (bool): Whether the search should be tautomer-insensitive.
        max_results (int): The maximum number of results to return.

    Returns:
        List of tuples containing (index, molecule, match) for each result.
    """
    bndl = rdMolEnumerator.Enumerate(query)

    # Added for safety
    if bndl.Size() == 0:
        bndl = [query]

    queries = []
    for q in bndl:
        if tautomer_insensitive:
            q = rdTautomerQuery.TautomerQuery(q)

        queries.append(q)


    matches = []
    for q in queries:
        matches.extend(sslib.GetMatches(q, params, maxResults=max_results))

    tmols = [(x, sslib.GetMol(x)) for x in matches]
    mols = []

    for idx, mol in sorted(tmols, key=lambda x: x[1].GetNumAtoms()):
        match = None
        for q in queries:
            if tautomer_insensitive:
                match = q.GetSubstructMatch(mol)
            else:
                match = mol.GetSubstructMatch(q)

        smiles_list = []
        smiles = Chem.MolToSmiles(mol)
        smiles_list.append(smiles)

        mols.append((idx, mol, smiles))

        if len(mols) >= max_results:
            break

    return mols


## Load the lib
sslib = load_latest_chembl()

## Set params
params = Chem.SubstructMatchParameters()
params.useGenericMatchers = True

## Perform the search
total_results = 0
for query_id, query_structure in queries:
    qry = Chem.MolFromSmarts(query_structure)
    Chem.SetGenericQueriesFromProperties(qry)

    smiles = []
    results = search_classes_library(qry, sslib, params)
    for match in results:
        s = match[2]
        smiles.append(s)

    print(f"{query_id} has {len(results)} results")

    for idx, smile in enumerate(smiles, start=1):
        print(f"Result {idx} for {query_id}: {smile}")

    total_results += len(results)

print(f"Total results across all queries: {total_results}")

```

**Comment by Adafede** [comment]
@baoilleach kindly suggested me to tag you @johnmay 👀

**Comment by johnmay** [comment]
A lot of this there is sometimes no obvious answer. Remember that CAS/SciFinder offer extensive training so people know how to use their markush search properly and there are sometimes multiple ways to do the same thing, i.e. link nodes vs SRU repeat etc - here be dragons.

# Structure 1-3

I agree, if there the degree is already constrained it makes sense to respect that. It's a likely tricky since it's ambiguous if ``*[#6,#7D3]*`` should have ``#6`` constrained to D2.

# Structure 4

One of the big "gotchas" is if you allow further extension or not. For example you specify the D<num> which is good... but you also need the system to understand attachments. This can be controlled with an option. In this case you could possibly  

- Only one ``[#6]`` can be attached, and nothing else.
``[#6]-[*].[c;D2]1[n][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n]1 |m:1:10.3|`` (as you had, not further attachment should be allowed)
- One, or two ``[#6]`` can be attached, but nothing else.
``[#6]-[*].[c;D2]1[n][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n]1 |m:1:3.10,Sg:n:0:1-2:ht|`` (as you had, multi attach only if allowed). Also 1,2 makes sense but you need to encode the comma.
- Only one ``[#6]`` can be attached, and if something else is attached on the other attachment point, it cannot be ``[#6]``.
``[#6]-[*].[c;D2]1[n][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n]1.[!#6]-[*] |m:1:3.10,12:3.10|``
perhaps
``[#6]-[*].[c;D2]1[n][c;D3]2[c;D2][c;D2][c;D2][n;D2][c;D3]2[n]1.[!#6]-[*] |m:1:3.10,12:3.10,Sg:n:11:0-1:ht|``

# Structure 5

Follows on from structure 4 but it's a big query. As you have written I would only allow one oxygen on each and nothing else.

# Structure 6

Never done to much on R group logic, so can't help here.

# Structure 7

I think links are always 1 implicitly. Note you can check what ChemAxon does https://marvinjs-demo.chemaxon.com/latest/, load in the example then save as normal SMARTS and it enumerates it for you.

You example does not work as expected, it loads it as 1-11-13, if you make it 1:1.13 then it does what I expect but not what you need here.

```
[CH3][CH2][N+0]([CH3])[CH3] |LN:1:11.13|
```

Personally I think it's reasonable to allow 0 as a repeat.

I would also expect.

```
[CH3][CH2][N+0]([CH3])[CH3] |Sg:n:1:11-13:ht|
```

to work equivaelently - it does not.

**Comment by johnmay** [comment]
If I get round to it I will add the link node quirk to by blog series:
- https://nextmovesoftware.com/blog/2022/04/08/cxsmiles-gotchas-part-1-bond-indexes/
- https://nextmovesoftware.com/blog/2022/04/20/cxsmiles-part-2-component-grouping/
